weave/integrations/langchain/langchain_test.py

```diff
@@ -0,0 +1,494 @@
+    from langchain_community.vectorstores import Chroma
+    from langchain_core.documents import Document
+    from langchain_core.output_parsers import StrOutputParser
+    from langchain_core.prompts import ChatPromptTemplate
+    from langchain_core.runnables import RunnablePassthrough
+    from langchain_openai import ChatOpenAI, OpenAIEmbeddings
+
+    loader = TextLoader("integrations/langchain/test_data/paul_graham_essay.txt")
+    docs = loader.load()
+
+    text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
+    splits = text_splitter.split_documents(docs)
+
+    api_key = os.environ.get("OPENAI_API_KEY", "sk-1234567890abcdef1234567890abcdef")
+
+    vectorstore = Chroma.from_documents(
+        documents=splits, embedding=OpenAIEmbeddings(openai_api_key=api_key)
+    )
+    retriever = vectorstore.as_retriever()
+
+    prompt = ChatPromptTemplate.from_template(
+        "You are an assistant for question-answering tasks. "
+        "Use the following pieces of retrieved context to answer the question. "
+        "If you don't know the answer, just say that you don't know. "
+        "Use three sentences maximum and keep the answer concise.\n"
+        "Question: {question}\nContext: {context}\nAnswer:"
+    )
+
+    llm = ChatOpenAI(
+        model_name="gpt-3.5-turbo", openai_api_key=api_key, temperature=0.0
+    )
+
+    def format_docs(documents: List[Document]) -> str:
+        return "\n\n".join(doc.page_content for doc in documents)
+
+    # Chain
+    rag_chain = (
+        {"context": retriever | format_docs, "question": RunnablePassthrough()}
+        | prompt
+        | llm
+        | StrOutputParser()
+    )
+
+    rag_chain.invoke(
+        input="What is the essay about?",
+    )
+
+    res = client.server.calls_query(tsi.CallsQueryReq(project_id=client._project_id()))
+    assert_correct_calls_for_rag_chain(res.calls)
+
+
+def assert_correct_calls_for_agent_with_tool(calls: list[tsi.CallSchema]) -> None:
+    assert len(calls) == 10
+
+    flattened = flatten_calls(calls)
+
+    got = [(op_name_from_ref(c.op_name), d) for (c, d) in flattened]
+
+    exp = [
+        ("langchain.Chain.AgentExecutor", 0),
+        ("langchain.Chain.RunnableSequence", 1),
+        ("langchain.Chain.RunnableParallel input chat_history agent_scratchpad ", 2),
+        ("langchain.Chain.RunnableLambda", 3),
+        ("langchain.Chain.RunnableLambda", 3),
+        ("langchain.Chain.RunnableLambda", 3),
+        ("langchain.Prompt.ChatPromptTemplate", 2),
+        ("langchain.Llm.ChatOpenAI", 2),
+        ("openai.chat.completions.create", 3),
+        ("langchain.Parser.OpenAIFunctionsAgentOutputParser", 2),
+    ]
+    assert got == exp
+
+
+@pytest.mark.skip_clickhouse_client
+@pytest.mark.vcr(
+    filter_headers=["authorization"],
+    allowed_hosts=["api.wandb.ai", "localhost", "trace.wandb.ai"],
+    before_record_request=filter_body,
+)
+def test_agent_run_with_tools(
```

Comment:
@ayulockin . Same here, Langchain uses streaming under the hood which causes the this test to fail.